modules/ubi/src/main/java/org/opensearch/ubi/ext/UbiParameters.java

```diff
@@ -0,0 +1,225 @@
+        PARSER.declareString(UbiParameters::setQueryId, QUERY_ID);
+        PARSER.declareString(UbiParameters::setUserQuery, USER_QUERY);
+        PARSER.declareString(UbiParameters::setClientId, CLIENT_ID);
+        PARSER.declareString(UbiParameters::setObjectId, OBJECT_ID);
+    }
+
+    private String queryId;
+    private String userQuery;
+    private String clientId;
+    private String objectId;
+
+    /**
+     * Get the {@link UbiParameters} from a {@link SearchRequest}.
+     * @param request A {@link SearchRequest},
+     * @return The UBI {@link UbiParameters parameters}.
+     */
+    public static UbiParameters getUbiParameters(final SearchRequest request) {
+
+        UbiParametersExtBuilder builder = null;
+
+        if (request.source() != null && request.source().ext() != null && !request.source().ext().isEmpty()) {
+            final Optional<SearchExtBuilder> b = request.source()
+                .ext()
+                .stream()
+                .filter(bldr -> UbiParametersExtBuilder.UBI_PARAMETER_NAME.equals(bldr.getWriteableName()))
+                .findFirst();
+            if (b.isPresent()) {
+                builder = (UbiParametersExtBuilder) b.get();
+            }
+        }
+
+        if (builder != null) {
+            return builder.getParams();
+        } else {
+            return null;
+        }
+
+    }
+
+    /**
+     * Creates a new instance.
+     */
+    public UbiParameters() {}
+
+    /**
+     * Creates a new instance.
+     * @param input The {@link StreamInput} to read parameters from.
+     * @throws IOException Thrown if the parameters cannot be read.
+     */
+    public UbiParameters(StreamInput input) throws IOException {
+        this.queryId = input.readString();
+        this.userQuery = input.readOptionalString();
+        this.clientId = input.readOptionalString();
+        this.objectId = input.readOptionalString();
+    }
+
+    /**
+     * Creates a new instance.
+     * @param queryId The query ID.
+     * @param userQuery The user-entered search query.
+     * @param clientId The client ID.
+     * @param objectId The object ID.
+     */
+    public UbiParameters(String queryId, String userQuery, String clientId, String objectId) {
+        this.queryId = queryId;
+        this.userQuery = userQuery;
+        this.clientId = clientId;
+        this.objectId = objectId;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder xContentBuilder, Params params) throws IOException {
+        return xContentBuilder.field(QUERY_ID.getPreferredName(), this.queryId)
+            .field(USER_QUERY.getPreferredName(), this.userQuery)
+            .field(CLIENT_ID.getPreferredName(), this.clientId)
+            .field(OBJECT_ID.getPreferredName(), this.objectId);
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(queryId);
```

Comment:
Should this be `getQueryId()` so that we always materialize the query ID before serializing?

Reply:
I think that is a good idea, especially for when the `query_id` is optional. Will update.